Libraries/LibCore/File.cpp

```diff
@@ -132,6 +134,30 @@ String File::real_path_for(const String& filename)
     return real_path;
 }
 
+bool File::ensure_parent_directories(const String& path)
+{
+    ASSERT(path[0] == '/');
+
+    char* parent_buffer = strdup(path.characters());
+    const char* parent = dirname(parent_buffer);
+
+    int rc = 0;
+    while (true) {
+        int rc = mkdir(parent, 0755);
+
+        if (rc == 0)
+            break;
+
+        if (errno != ENOENT)
+            break;
+
+        ensure_parent_directories(parent);
+    };
+
+    free(parent_buffer);
+    return rc == 0;
```

Comment:
By this time, `errno` is going to be changed by the `free()` call.

Reply:
Any why does it matter? `rc` only tracks the return code of `mkdir`.

Reply:
`rc` yes, but `errno` tracks the last call. If `mkdir()` fails, you'd return `false` form here, and then the caller would see that and use something like `perror("ensure_parent_directories")`, but would get the wrong error message.

Reply:
Do we expect `perror` to work with non-`LibC` functions?
For example, it would make using `AK` containers in LibCore functions problematic since they could call `free` in their destructor.

Reply:
That depends on the API in question, but generally, if it's fallible, it should use one of the following styles:
* return `-1`/`nullptr`/`false` and set `errno`
* do the same and store the error somewhere inside the object; this is what `Core::IODevice` and stdio `FILE` do:
    https://github.com/SerenityOS/serenity/blob/9225bfa95e354dc12c6705d65ff33224095a8917/Libraries/LibCore/IODevice.h#L103
    https://github.com/SerenityOS/serenity/blob/9225bfa95e354dc12c6705d65ff33224095a8917/Libraries/LibC/stdio.cpp#L135
* return a `Result<>`

`perror()` only works with `errno`, so it's only suitable after a call to an API that sets `errno` (though you can set `errno` yourself if you're using an API that gives you the error in a different way).

Reply:
Okay, `ensure_parent_directories` should now return with an errno that reflects the reason for the failure. Let me know what you think.

Reply:
This looks pretty good, assuming C++ guarantees that the destruction order is the reverse of the creation order.